syntax = "proto3";

package tripsphere.chat;

import "google/protobuf/timestamp.proto";
import "tripsphere/chat/memory.proto";

enum MessageRole {
  MESSAGE_ROLE_USER = 0;
  MESSAGE_ROLE_AGENT = 1;
  MESSAGE_ROLE_ASSISTANT = 2;
  MESSAGE_ROLE_SYSTEM = 3;
}

enum PartType {
  PART_TYPE_TEXT = 0;
  PART_TYPE_FILE = 1;
  PART_TYPE_DATA = 2;
}

message Part {
  PartType type = 1;
  bytes content = 2;
  map<string, string> metadata = 3;
}

message Message {
  string message_id = 1;
  MessageRole role = 3;
  repeated Part parts = 4;
}

message Session {
  string session_id = 1; // ID of the session
  string user_id = 2; // ID of the user who owns the session
  MemoryStrategy memory_strategy = 4; // Memory strategy for the session
  google.protobuf.Timestamp created_at = 5;
  google.protobuf.Timestamp updated_at = 6;
}

service SessionService {
  // Create a new chat session.
  rpc CreateSession(CreateSessionRequest) returns (CreateSessionResponse) {}

  // Delete a chat session.
  rpc DeleteSession(DeleteSessionRequest) returns (DeleteSessionResponse) {}

  // Get details of a chat session.
  rpc GetSession(GetSessionRequest) returns (GetSessionResponse) {}

  // Get messages of a chat session.
  rpc GetSessionMessages(GetSessionMessagesRequest) returns (GetSessionMessagesResponse) {}
}

message CreateSessionRequest {
  string user_id = 1; // ID of the user creating the session
  MemoryStrategy memory_strategy = 2; // Memory strategy for the session
}

message CreateSessionResponse {
  Session session = 1; // Details of the created session
}

message DeleteSessionRequest {
  string session_id = 1; // ID of the session to delete
}

message DeleteSessionResponse {}

message GetSessionRequest {
  string session_id = 1; // ID of the session to retrieve
}

message GetSessionResponse {
  Session session = 1; // Details of the requested session
}

enum Direction {
  DIRECTION_FORWARD = 0;
  DIRECTION_BACKWARD = 1;
}

message GetSessionMessagesRequest {
  string session_id = 1; // ID of the session to retrieve messages from
  int32 limit = 2; // Maximum number of messages to retrieve
  int32 offset = 3; // Offset for pagination
  Direction direction = 4; // Direction of message retrieval
}

message GetSessionMessagesResponse {
  repeated Message messages = 1; // List of messages in the session
}